lib/ImportFromC/HeaderToModel.cpp

```diff
@@ -1144,36 +1038,33 @@ bool DeclVisitor::VisitEnumDecl(const EnumDecl *D) {
   auto TheUnderlyingModelType = getEnumUnderlyingType(*UnderlyingType);
   if (not TheUnderlyingModelType) {
     revng_log(Log,
-              "UnderlyingType of a EnumType can only be Signed or Unsigned");
+              "UnderlyingType of a EnumDefinition can only be Signed or "
+              "Unsigned");
     return false;
   }
 
-  auto NewType = makeType<model::EnumType>();
-  if (AnalysisOption == ImportFromCOption::EditType)
-    NewType->ID() = (*Type)->ID();
-
-  auto *Definition = D->getDefinition();
-  auto TypeEnum = cast<model::EnumType>(NewType.get());
-  model::QualifiedType TheUnderlyingType(*TheUnderlyingModelType, {});
-  TypeEnum->UnderlyingType() = TheUnderlyingType;
-  setCustomName(*TypeEnum, Definition->getName());
-
-  for (const auto *Enum : Definition->enumerators()) {
-    auto &EnumEntry = TypeEnum->Entries()[Enum->getInitVal().getExtValue()];
-    std::string NewName = Enum->getName().str();
-    if (TypeEnum->entryName(EnumEntry) != NewName)
-      EnumEntry.CustomName() = NewName;
+  model::EnumDefinition *NewType = nullptr;
+  if (AnalysisOption == ImportFromCOption::EditType) {
+    revng_assert(Type != std::nullopt);
+    model::TypeDefinition &Definition = *Model->TypeDefinitions().at(*Type);
+    if (auto *Enum = llvm::dyn_cast<model::EnumDefinition>(&Definition)) {
+      NewType = Enum;
+    } else {
+      // It seems like the kind of the type got changed. Since it affects
+      // the key we need to erase the old type before adding the new one.
+      Model->TypeDefinitions().erase(*Type);
+      NewType = &Model->makeEnumDefinition().first;
```

Comment:
So this is something you added, that wasn't there before this PR.

What's the case you're trying to handle? And why haven't you handled that case for other `TypeDefinitions`?

Reply:
It was still there, just at the very bottom. The only change is that I'm erasing the type directly instead of using `std::erase_if`.

Reply:
> And why haven't you handled that case for other `TypeDefinitions`?

Other alternatives can fail. Which makes this reordering a bit tricky for them. The enum one always succeeds, so we can get away with creating the type first. In the perfect world, we'd use `TypeBucket` for the others to make them similar. But I deemed that to much, got the low hanging enum fruit and went on.

Reply:
Allright.